activerecord/lib/active_record/relation/finder_methods.rb

```diff
@@ -145,15 +145,23 @@ def first!
     #
     #   [#<Person id:4>, #<Person id:3>, #<Person id:2>]
     def last(limit = nil)
-      if limit
-        if order_values.empty? && primary_key
-          order(arel_table[primary_key].desc).limit(limit).reverse
-        else
-          to_a.last(limit)
-        end
-      else
-        find_last
-      end
+      return find_last(limit) if loaded?
+
+      result = order_values.empty? && primary_key ? order(arel_table[primary_key].desc) : reverse_order
+      result = result.limit!(limit || 1)
+      limit ? result.reverse : result.first
+    rescue ActiveRecord::IrreversibleOrderError
+      ActiveSupport::Deprecation.warn(<<-WARNING.squish)
+          Finding a last element by loading the relation when SQL ORDER
+          can not be reversed is deprecated.
+          Rails 5.1 will raise ActiveRecord::IrreversibleOrderError in this case.
+          Please call `to_a.last` if you still want to load the relation.
+      WARNING
```

Comment:
Saw you raised `IrreversibleOrderError` with two different messages in https://github.com/rails/rails/pull/18928. At least the first one gives you more info than this warning does alone.

On the other hand, you do specify that people should just add `to_a` which I guess would be the fix anyway. So I'll backtrack: this is fine as is 👍